src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -7134,14 +7136,179 @@ std::pair<MergeTreeData::MutableDataPartPtr, scope_guard> MergeTreeData::cloneAn
     if (params.copy_instead_of_hardlink)
         with_copy = " (copying data)";
 
-    auto dst_part_storage = src_part_storage->freeze(
-        relative_data_path,
-        tmp_dst_part_name,
-        read_settings,
-        write_settings,
-        /* save_metadata_callback= */ {},
-        params);
+    std::shared_ptr<IDataPartStorage> dst_part_storage{};
+    try
+    {
+        dst_part_storage = src_part_storage->freeze(
+            relative_data_path,
+            tmp_dst_part_name,
+            read_settings,
+            write_settings,
+            /* save_metadata_callback= */ {},
+            params);
+    }
+    catch (...)
+    {
+        /// Hardlink fail. Try copy.
+        LOG_WARNING(
+            &Poco::Logger::get("MergeTreeData"),
+            "Hard link fail, try tp copy directly. to:{}, path:{}",
+            this->getRelativeDataPath(),
+            tmp_dst_part_name);
+        bool copy_successful = false;
+        for (const DiskPtr & disk : this->getStoragePolicy()->getDisks())
+        {
+            try
+            {
+                auto reservation_space = src_part_storage->reserve(src_part->getBytesOnDisk());
+                if (!reservation_space)
+                    throw Exception(ErrorCodes::NOT_ENOUGH_SPACE, "Not enough space on disk.");
+                dst_part_storage = src_part_storage->clonePart(
+                    this->getRelativeDataPath(), tmp_dst_part_name, disk, read_settings, write_settings, {}, {});
+                copy_successful = true;
+                break;
+            }
+            catch (Exception & e)
+            {
+                LOG_TRACE(&Poco::Logger::get("MergeTreeData"), "Clone part on disk {} fail: {}", disk->getName(), e.what());
+            }
+        }
+        if (!copy_successful)
+            LOG_FATAL(&Poco::Logger::get("MergeTreeData"), "Hard link fail, clone fail");
```

Comment:
It isn't fatal.

Reply:
Both copy and hard link failed, indicating that the attaching partition was not successful. I change the level to "Error".